journalbeat/input/input.go

```diff
@@ -0,0 +1,141 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package input
+
+import (
+	"github.com/elastic/beats/journalbeat/checkpoint"
+	"github.com/elastic/beats/journalbeat/reader"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// Input manages readers and forwards entries from journals.
+type Input struct {
+	readers  []*reader.Reader
+	done     chan struct{}
+	config   Config
+	pipeline beat.Pipeline
+	states   map[string]checkpoint.JournalState
+}
+
+// New returns a new Inout
+func New(
+	c *common.Config,
+	pipeline beat.Pipeline,
+	done chan struct{},
+	states map[string]checkpoint.JournalState,
+) *Input {
+	config := DefaultConfig
+	if err := c.Unpack(&config); err != nil {
+		logp.Err("Error unpacking config: %v", err)
+		return nil
+	}
+	var readers []*reader.Reader
+	if len(config.Paths) == 0 {
+		cfg := reader.Config{
+			Path:          reader.LocalSystemJournalID, // used to identify the state in the registry
+			Backoff:       config.Backoff,
+			MaxBackoff:    config.MaxBackoff,
+			BackoffFactor: config.BackoffFactor,
+			Seek:          config.Seek,
+		}
+
+		state := states[reader.LocalSystemJournalID]
+		r, err := reader.NewLocal(cfg, done, state)
+		if err != nil {
+			logp.Err("Error creating reader for local journal: %v", err)
+			return nil
+		}
+		readers = append(readers, r)
+	}
+
+	for _, p := range config.Paths {
+		cfg := reader.Config{
+			Path:          p,
+			Backoff:       config.Backoff,
+			MaxBackoff:    config.MaxBackoff,
+			BackoffFactor: config.BackoffFactor,
+			Seek:          config.Seek,
+		}
+		state := states[p]
+		r, err := reader.New(cfg, done, state)
+		if err != nil {
+			logp.Err("Error creating reader for journal: %v", err)
+			continue
+		}
+		readers = append(readers, r)
+	}
+
+	return &Input{
+		readers:  readers,
+		done:     done,
+		config:   config,
+		pipeline: pipeline,
+		states:   states,
+	}
+}
+
+// Run connects to the output, collects entries from the readers
+// and then publishes the events.
+func (i *Input) Run() {
+	if len(i.readers) == 0 {
+		return
+	}
+
+	client, err := i.pipeline.ConnectWith(beat.ClientConfig{
+		PublishMode:   beat.GuaranteedSend,
+		EventMetadata: common.EventMetadata{},
+		Meta:          nil,
+		Processor:     nil,
+		ACKCount: func(n int) {
+			logp.Info("journalbeat successfully published %d events", n)
+		},
+	})
+	if err != nil {
+		logp.Err("Error connecting: %v", err)
+		return
+	}
+	defer client.Close()
+
+	for {
+		select {
+		case <-i.done:
+			return
+		default:
+			for _, r := range i.readers {
+				for e := range r.Follow() {
+					client.Publish(*e)
+				}
+			}
+		}
+	}
```

Comment:
I am not sure the above is doing what you are expecting. Adding the `Follow` implementation in the comment, Note I need to read more on the Reader class but at first look we could have an issue.

```
func (r *Reader) Follow() <-chan *beat.Event {
	out := make(chan *beat.Event)
	go r.readEntriesFromJournal(out)
	return out
}

```

The above return a channel that we consume with range, this means we will consume all the events until events are available or the channel is closed, looking at the reader implementation we only close the channel when we received a done signal.

Since the publish to the queue is done in a single goroutine, it is possible that a specific reader in the list make all the other readers starves for resource and the possibility to push publish events?

Reply:
I think we could fix it in two differents way:


- Provide a call back on new events and that callback will do the actual push to the queue.
- Merge N goroutines into a single goroutine and read from that. Similar to [this link](https://medium.com/justforfunc/two-ways-of-merging-n-channels-in-go-43c0b57cd1de)

Reply:
Now I tested it again for multiple runs. It seems like sometimes harvesters are starved. I haven't noticed it until you pointed it out. Thanks!